### PR TITLE
implement browser clipboard service

### PR DESCRIPTION
**Pull request by sbatten**

### CODE REVIEW

File: src/vs/workbench/contrib/files/browser/fileActions.ts [hunk 1]
```diff
@@ -836,8 +836,8 @@ class ClipboardContentProvider implements ITextModelContentProvider {
 		@IModelService private readonly modelService: IModelService
 	) { }
 
-	provideTextContent(resource: URI): Promise<ITextModel> {
-		const model = this.modelService.createModel(this.clipboardService.readText(), this.modeService.createByFilepathOrFirstLine(resource), resource);
+	async provideTextContent(resource: URI): Promise<ITextModel> {
+		const model = this.modelService.createModel(await this.clipboardService.readText(), this.modeService.createByFilepathOrFirstLine(resource), resource);
 
 		return Promise.resolve(model);
```
Comment (bpasero) [inline src/vs/workbench/contrib/files/browser/fileActions.ts]:
```suggestion
		return model;
```

File: src/vs/platform/clipboard/browser/clipboardService.ts [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { registerSingleton } from 'vs/platform/instantiation/common/extensions';
+import { IClipboardService } from 'vs/platform/clipboard/common/clipboardService';
+import { URI } from 'vs/base/common/uri';
+
+export class ClipboardService implements IClipboardService {
+
+	_serviceBrand: any;
```
Comment (bpasero) [inline src/vs/platform/clipboard/browser/clipboardService.ts]:
```suggestion
	_serviceBrand: ServiceIdentifier<IClipboardService>;
```

File: src/vs/editor/contrib/snippet/snippetVariables.ts [hunk 3]
```diff
@@ -49,7 +49,7 @@ export class CompositeSnippetVariableResolver implements VariableResolver {
 		//
 	}
 
-	resolve(variable: Variable): string | undefined {
+	async resolve(variable: Variable): Promise<string | undefined> {
 		for (const delegate of this._delegates) {
 			let value = delegate.resolve(variable);
 			if (value !== undefined) {
```
Comment (bpasero) [inline src/vs/editor/contrib/snippet/snippetVariables.ts]: @sbatten this is broken now, you have to `await` the value, no?

**Review by bpasero — CHANGES_REQUESTED** [review]
@sbatten can you please revisit where you changed methods to return a `Promise` and forgot to add an `await`, I found at least https://github.com/microsoft/vscode/pull/75293/files#diff-058cba421d6dc4a91f615f8a9e35adb8R553

Also, are we certain `navigator.clipboard` works as advertised? I think it can be `undefined` in certain cases so we might need to probe for that and install a workaround (see https://stackoverflow.com/questions/51805395/navigator-clipboard-is-undefined). 

https://caniuse.com/#search=clipboard seems to indicate that the clipboard API is only available on user interactions (e.g. click), though not sure that still holds true. 

Maybe @alexandrudima and @rebornix could help out here, given we seem to have some way of clipboard support (via https://github.com/Microsoft/vscode/blob/39947c72507deb8a44de785825243b9da30a7a19/src/vs/editor/contrib/clipboard/clipboard.ts#L22) in the form of `document.execCommand`).

**Comment by rebornix** [comment]
The last time when I modified Monaco’s clipboard service related code, we can access the clipboard only in events handler of cut, copy and paste but not navigator.clipboard . @sbatten we can take a look together.

**Comment by rebornix** [comment]
Re https://developer.mozilla.org/en-US/docs/Web/API/Clipboard

> Calls to the methods of the Clipboard object will not succeed if the user hasn't granted the needed permissions using the Permissions API and the "clipboard-read" or "clipboard-write" permission as appropriate.

I'm wondering if we should use `document.execCommand` by default if we can do that (say there is a `copy` event handler registered for the focused element, or it's an input/textarea/etc), and then fall back to `navigator.clipboard`. We don't want to *bother* with permission access when it's not necessary.

File: src/vs/workbench/contrib/debug/browser/debugActions.ts [hunk 4]
```diff
@@ -396,11 +396,11 @@ export class CopyValueAction extends Action {
 
 		if (this.value instanceof Variable && stackFrame && session && this.value.evaluateName) {
 			return session.evaluate(this.value.evaluateName, stackFrame.frameId, this.context).then(result => {
-				this.clipboardService.writeText(result.body.result);
-			}, err => this.clipboardService.writeText(this.value.value));
+				return this.clipboardService.writeText(result.body.result);
+			}, err => { return this.clipboardService.writeText(this.value.value); });
```
Comment (roblourens) [inline src/vs/workbench/contrib/debug/browser/debugActions.ts]: The `err` change is redundant, does the same as the original.

File: src/vs/workbench/contrib/debug/browser/repl.ts [hunk 5]
```diff
@@ -470,16 +470,15 @@ export class Repl extends Panel implements IPrivateReplService, IHistoryNavigati
...
 			if (nativeSelection) {
-				this.clipboardService.writeText(nativeSelection.toString());
+				await this.clipboardService.writeText(nativeSelection.toString());
 			}
 			return Promise.resolve();
 		}));
 		actions.push(new Action('workbench.debug.action.copyAll', nls.localize('copyAll', "Copy All"), undefined, true, () => {
-			this.clipboardService.writeText(this.getVisibleContent());
-			return Promise.resolve(undefined);
+			return Promise.resolve(this.clipboardService.writeText(this.getVisibleContent()));
```
Comment (roblourens) [inline src/vs/workbench/contrib/debug/browser/repl.ts]: Seems like we should use async vs promises consistently in the same file at least, the change above this uses async

File: src/vs/workbench/contrib/preferences/browser/keybindingsEditor.ts [hunk 6]
```diff
@@ -709,7 +709,7 @@ export class KeybindingsEditor extends BaseEditor implements IKeybindingsEditor
 			label: localize('copyLabel', "Copy"),
 			enabled: true,
 			id: KEYBINDINGS_EDITOR_COMMAND_COPY,
-			run: () => this.copyKeybinding(keybindingItem)
+			run: () => { return this.copyKeybinding(keybindingItem); }
```
Comment (roblourens) [inline src/vs/workbench/contrib/preferences/browser/keybindingsEditor.ts]: Same as above, `() => foo` returns foo

**Comment by sbatten** [comment]
@bpasero updated with feedback. The link for caniuse is referring to the older synchronous clipboard API. The new API is async but not well supported across browsers. It does what we need which is a method to read and write arbitrary text to the clipboard. After discussing with @rebornix, this is what the service is responsible for while general copy, cut, paste are handled separately using the synchronous API. For me, this means our service does the best it can with the async API and throws when the browser does not support this. We cannot fallback to the synchronous API since there is no meaning to copy text that isn't selected when an extension asks to write text to the clipboard.

**Review by bpasero — CHANGES_REQUESTED** [review]
@sbatten I mentioned this before, you changed methods to return a `Promise` but you are not `awaiting` them properly. E.g. `Variable.resolve()`:

```ts
resolveVariables(resolver: VariableResolver): this {
	this.walk(candidate => {
		if (candidate instanceof Variable) {
			if (candidate.resolve(resolver)) {
				this._placeholders = undefined;
			}
		}
		return true;
	});
	return this;
}
```

**Comment by sbatten** [comment]
@bpasero yea i thought i got all these with https://github.com/microsoft/vscode/pull/75293/commits/37e5c97708e6e7435fe24fc516f959473517f22b but as I continue, this is exploding more and more. The assumption of a synchronous clipboard is very pervasive. If I continue this way, I am going to have to write some tooling to do the conversion as it is becoming very difficult to keep track of what's changed and what needs to change. It does seem that the explosion roots from the clipboard based snippets however, so maybe there is a way to reduce the impact by modifying that a bit.

**Comment by bpasero** [comment]
I am not sure I can follow. There shouldn't be a ton of users of clipboard service, or? And I would assume that in many cases the usage of the clipboard service is one of the last things in the code (e.g. as a result of an action), so doing an `await...` should be no problem?

**Comment by sbatten** [comment]
@bpasero yes clipboard service usage is low but once that new location's function becomes async, this permeates, and where it is changed, the value is actually needed from readText. I'm looking into splitting some function calls to async and sync versions to limit the impact to the primary code path that is the problem (clipboard based snippets)

**Comment by bpasero** [comment]
@sbatten oh now I see what you mean: if someone calls `writeText` and someone `readText` without waiting, you might end up reading stale values?

I think the only solution here is to add a [`Barrier`](https://github.com/Microsoft/vscode/blob/f3fd9f71b6480ad35340b8835315212a4481bea3/src/vs/base/common/async.ts#L287) that synchronizes read/write access. We have utility classes for this in `async.ts`

**Comment by sbatten** [comment]
@bpasero while that may also be an issue, it wasn't exactly what I am referring to. 

I'm referring to the following:

```
function A() {
 const val = B();
 return val !== undefined;
}

function B() {
 return true;
}
```

In the above example, imagine `B` becomes `async`. The same would need to happen to `A` because it needs to `await` the value of `B`. `B` may only have one reference in the whole code base. But `A` has *N*, *C_1* ... *C_N*. If we iterate on this process, it explodes quickly as *C_x* can also have may references. Further, even with a linting rule applied, this case is not detected because we store the promise of `B` so we don't know it is floating. So, finding all references and updating has been the approach so far but like I said, one changes brings 10 others. Eventually this will end, but its a dangerous change. The result is me looking into a smaller more local refactor of snippets or tooling to make sure the larger change is safe.

**Comment by bpasero** [comment]
@sbatten yes both things are valid:
* we need to sync on reading and writing
* we need to change all methods that read/write clipboard to become async at least if the caller relies on the result of the call. I think we can assume that this is only the case for reading but not writing

**Comment by sbatten** [comment]
@bpasero 

>we need to change all methods that read/write clipboard to become async at least if the caller relies on the result of the call. I think we can assume that this is only the case for reading but not writing

agree and since this is becoming a dangerously large and unvalidate-able change, I want to minimize it or use some smarter tooling to detect issues. for context I have another incomplete local commit with another 10-20 files that have to change on top of what's here.

**Comment by bpasero** [comment]
Ok, just add it and I can review again.

File: src/vs/editor/contrib/suggest/suggestController.ts [hunk 7]
```diff
@@ -423,7 +423,7 @@ const SuggestCommand = EditorCommand.bindToContribution<SuggestController>(Sugge
 registerEditorCommand(new SuggestCommand({
 	id: 'acceptSelectedSuggestion',
 	precondition: SuggestContext.Visible,
-	handler: x => x.acceptSelectedSuggestion(true),
+	handler: async (x) => { await x.acceptSelectedSuggestion(true); },
```
Comment (roblourens) [inline src/vs/editor/contrib/suggest/suggestController.ts]: Functions like this don't need to use async/await, the fat arrow function is already returning the promise

Comment (roblourens) [reply src/vs/editor/contrib/suggest/suggestController.ts]: But it's also fine to write it this way

Comment (sbatten) [reply src/vs/editor/contrib/suggest/suggestController.ts]: @roblourens Yea you're right. Originally, I was planning to make them all with async/await for clarity reasons, but I'm not sure I was consistent in that approach.

**Comment by sbatten** [comment]
@bpasero finished all updates... I hope. Ran unit tests to ensure no changes, caught one missed await in the test itself.

**Comment by bpasero** [comment]
Adding @jrieken for changes in snippets land.

File: src/vs/editor/contrib/snippet/snippetController2.ts [hunk 8]
```diff
@@ -60,17 +60,17 @@ export class SnippetController2 implements IEditorContribution {
 		return 'snippetController2';
 	}
 
-	insert(
+	async insert(
```
Comment (jrieken) [inline src/vs/editor/contrib/snippet/snippetController2.ts]: 🛑 this cannot be async, never. Subsequently all following up async changes cannot happen. 

Making snippets insert async, means that insert suggestion become async and that means you can accept a suggestion, type, and get the suggestion insert after that. This needs a different approach, e.g the snippet completion provider has the change to read the clipboard at the time of computing snippets (async is OK there)

**Review by jrieken — CHANGES_REQUESTED** [review]
Sorry, this needs a different approach for snippets et al

**Review by bpasero — APPROVED** [review]
@sbatten a couple of feedback (outside of snippets/terminal):
- you do not need to ever use `Promise.resolve()` if the thing is a promise you return, I noticed you are using it in a couple of places in your change
- `startupProfiler`: suggest to convert this to use `async `and not a mix of `async+then`

**Review by Tyriar — APPROVED** [review]
Terminal 👌

**Comment by sbatten** [comment]
@jrieken updated with your changes to remove nearly all changes to snippets code. please verify

@bpasero updated to address your feedback

**Review by jrieken — APPROVED** [review]

**Pull request merged**